keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,57 +201,340 @@ ensure all tenants have rights to the images that are already present on a host.
+### Kubelet Caching
+
+The kubelet tracks a pulled image auth cache for the credentials that were successfully used to pull an image.
+This also includes anonymous pulls and pulls with node-wide credentials. This cache will be persisted to disk using 
+well-defined structure that follows standard config file API versioning techniques.
+This is done to keep track of the authenticated images between kubelet restarts.
+
+A cache entry for an image will be deleted when the image is garbage collected. Thus, the max size of the
+cache scales with the number of unique cache entries * the number of unique images that have not been garbage collected.
+It is not expected that this would be a significant number of bytes. This claim shall be verified by actual use in Alpha and subsequent metrics in Beta.
+
+For Alpha, the cache should be implemented solely in files and then be duplicated to
+memory in Beta.
+
+#### Caching Policies
+
+Some images that already exist on the kubelet's node and are not tracked might be ignored
+for credential reverification (and therefore ignored for caching).
+
+The rules that define what should be ignored are set by the `imagePullCredentialsVerificationPolicy`:
+- `NeverVerify` matches the previous behavior from user perspective. Kubelet still creates
+   records of image pulls as these would be required should the policy change later.
+- `NeverVerifyPreloadedImages` images pulled without a secret and those pulled
+  outside the kubelet process are ignored (new default behavior)
+- `NeverVerifyAllowlistedImages` - images pulled without a secret and those pulled
+  outside the kubelet process but mentioned in `preloadedImagesVerificationAllowlist` are ignored
+- `AlwaysVerify` - all the image pulls need reverification.
+
+#### Writing to the Cache
+
+There are two different stages when the kubelet updates its cache:
+1. **Before the image pull**
+    - An image pull intent record is written befor the actual pull occurs.
+    - This is is to mitigate a situation where the kubelet crashes during the pull and an
+      image that gets pulled in the meantime later appears as if it were *preloaded*.
+2. **After the pull**
+    - Once we know the result of the pull operation, we either remove the image pull
+      intent record (=> failed pull attempt), or we write a new record of the image
+      being pulled successfully and remove the record of the pull intent.
+    - Without this, if kubelet failed during an image pull, an existing record would
+      already grant image access irregardless of this pull operation result for any next
+      pull attempts.
+
+The kubelet uses the following Golang interface to interact with the cache:
+```go
+// ImagePullCredentials wraps credentials that can be used to pull an image.
+// One of the attributes must always be non-nil, and the others must be nil.
+type ImagePullCredentials struct {
+    KubernetesSecret *imagemanager.ImagePullSecret
+    NodeWideCredentials *bool
+    Anonymous *bool
```

Comment:
can collapse these to a single field named something like `AllowAllPods *bool`

Reply:
I started using `NodePodsAccessible` but I don't insist on it if `AllowAllPods` or any other feels more appropriate to the native speaker's ear (eye?).